### PR TITLE
AVRO-2090: Improve encode/decode time for SpecificRecord using code generation

### PR DESCRIPTION
**Pull request by rstata**
Initial patch for AVRO-2090

### CODE REVIEW

**Comment by scottcarey** [comment]
This is a good idea.  What is the performance improvement?  Did you run any benchmarks?

I had imagined generating bytecode using ASM, but generating it in the generated class would work too.

File: lang/java/avro/src/main/java/org/apache/avro/specific/SpecificData.java [hunk 1]
```diff
@@ -122,6 +122,10 @@ public DatumWriter createDatumWriter(Schema schema) {
   /** Return the singleton instance. */
   public static SpecificData get() { return INSTANCE; }
 
+  private static final boolean USE_CUSTOM_CODERS
+    = Boolean.parseBoolean(System.getProperty("org.apache.avro.specific.use_custom_coders","false"));
+  public boolean useCustomCoders() { return USE_CUSTOM_CODERS; }
```
Comment (cutting) [inline lang/java/avro/src/main/java/org/apache/avro/specific/SpecificData.java]: useCustomCoders should probably be a settable member variable on each SpecificData instance, so folks can more easily switch it on and off without restarting the JVM.

File: lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumWriter.java [hunk 2]
```diff
@@ -71,6 +71,21 @@ protected void writeString(Schema schema, Object datum, Encoder out)
     writeString(datum, out);
   }
 
+  @Override
+  protected void writeRecord(Schema schema, Object datum, Encoder out)
+    throws IOException {
+    if (SpecificData.get().useCustomCoders()
```
Comment (cutting) [inline lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumWriter.java]: should use this.getSpecificData() instead of the static instance

**Comment by rstata** [comment]
This work has gotten a bit old in my head.  I can dust it off and address the issues raised in this ticket, but want to make sure there is genuine interest before making the effort.

**Comment by rstata** [comment]
BTW, I believe that Thiru presented a bit of this work at Strata last Fall:

https://conferences.oreilly.com/strata/strata-ny-2017/public/schedule/detail/60729

**Comment by cutting** [comment]
If it demonstrates a big performance improvement under Perf.java then I think it would be a good change to have & am willing to help get it committed.

**Comment by juwex** [comment]
I'm very excited about the performance improvements with this feature and would like to see this in a usable state sooner rather than later. If I can assist getting this branch into a mergeable state, I'd love to do so (with @rstata 's consent)

There's a rather big issue with schema evolution currently, however. The generated reader relies on the field order within the `ResolvingDecoder` to stay the same. Hence, data written with a schema that only differs in regards to the order of fields will result in an error or (even worse) invalid data. If required, I can provide a code sample.

File: lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java [hunk 3]
```diff
@@ -101,6 +101,23 @@ private Class getPropAsClass(Schema schema, String prop) {
     }
   }
 
+  @Override
+  protected Object readRecord(Object old, Schema expected, ResolvingDecoder in)
+    throws IOException {
+    SpecificData data = getSpecificData();
+    Object r = data.newRecord(old, expected);
```
[the remark this answers is not in the document]
Comment (rstata) [reply lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java]: yep

File: lang/java/avro/src/main/java/org/apache/avro/specific/SpecificRecordBase.java [hunk 4]
```diff
@@ -90,4 +92,19 @@ public void readExternal(ObjectInput in)
...
       .read(this, SpecificData.getDecoder(in));
   }
+
+  /** Returns true iff an instance supports the {@link #encode} and
+    * {@link #decode} operations.  Should only be used by
+    * <code>SpecificDatumReader/Writer</code> to selectively use
+    * {@link #encode} and {@link #decode} to optimize the (de)serialization of
+    * values. */
+  public boolean hasCustomCoders() { return false; }
+
```
[the remark this answers is not in the document]
Comment (rstata) [reply lang/java/avro/src/main/java/org/apache/avro/specific/SpecificRecordBase.java]: protected is enough

File: lang/java/compiler/src/main/velocity/org/apache/avro/compiler/specific/templates/java/classic/record.vm [hunk 5]
```diff
@@ -473,4 +475,282 @@ public class ${this.mangle($schema.getName())}#if ($schema.isError()) extends or
...
+#end
+
+#macro( decodeArray $indent $var $s $rs )
+#set ($I = $this.indent($indent))
+#set ($t = $this.javaType($s))
+#set ($et = $this.javaType($s.ElementType))
+#set ($gat = "SpecificData.Array<${et}>")
+$I    long size${nv} = in.readArrayStart();
+$I    $t a${nv} = ${var}; // Need fresh name due to limitation of macro system
+$I    if (a${nv} == null) {
```
[the remark this answers is not in the document]
Comment (rstata) [reply lang/java/compiler/src/main/velocity/org/apache/avro/compiler/specific/templates/java/classic/record.vm]: changed to a velocity comment

File: lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumWriter.java (hunk 2 above)
Comment (rstata) [reply lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumWriter.java]: ok

File: lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java [hunk 6]
```diff
@@ -101,6 +101,23 @@ private Class getPropAsClass(Schema schema, String prop) {
...
+    if (SpecificData.get().useCustomCoders()
```
[the remark this answers is not in the document]
Comment (rstata) [reply lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java]: ok

**Comment by rstata** [comment]
I will update this patch.  The old changes were easily rebased onto Avro's current master.  You can see that result here:

https://github.com/rstata-projects/avro/tree/specific-new-rebased

I then started addressing the comments above.  I just closed a few of the change requests from above -- you can find the resulting changes here:

https://github.com/rstata-projects/avro/tree/AVRO-2090-again

I'll keep working on this over the next few days.  In the meantime, if folks have applications that use SpecificRecord, it'd be great if you could enable this feature and run it through your test cases.

**Comment by rstata** [comment]
@juwex, would love your help if you're still interested.  Good catch on the field-order problem.  It'd be great if you could develop a test case that demonstrates the bug you've found -- it'd be a fantastic regression test.  Regarding a fix for this problem, we should check to see if the Decoder provided is an instance of ResolvingDecoder, and if it is code should use readFieldOrder to read fields in the correct order.

File: lang/java/avro/src/main/java/org/apache/avro/specific/SpecificData.java (hunk 1 above)
Comment (rstata) [reply lang/java/avro/src/main/java/org/apache/avro/specific/SpecificData.java]: In https://github.com/rstata-projects/avro/commit/e81a5880be69aef5bffdcf38f7ae3491957a6a68, I changed this flag to be dynamically changable.

For the record, here's the roll-out plan I imagined for this feature (See [AVRO-2091](https://issues.apache.org/jira/browse/AVRO-2091)).  The initial release of this feature sets the `use_custom_coders` flag to `false` by default, to allow for some road testing.  The  release after that would set this flag to `true` by default, to allow for some more serious road testing.  The next release of Avro after that would then eliminate this flag altogether, under the assumption that the feature has been sufficiently tested and there's no longer a use for the old way of doing things.

**Comment by rstata** [comment]
I've opened a new pull request (#350) to replace this one.

**Pull request closed without merging**